packages/calcite-components/src/components/carousel/carousel.tsx

```diff
@@ -609,31 +646,52 @@ export class Carousel
     </div>
   );
 
-  renderPaginationItems = (): VNode => (
-    <div aria-label={this.label} class={CSS.paginationItems} role="tablist">
-      {this.items.map((item, index) => {
-        const isMatch = index === this.selectedIndex;
-        return (
-          <button
-            aria-controls={!isMatch ? item.id : undefined}
-            aria-selected={toAriaBoolean(isMatch)}
-            class={{
-              [CSS.paginationItem]: true,
-              [CSS.paginationItemIndividual]: true,
-              [CSS.paginationItemSelected]: isMatch,
-            }}
-            data-index={index}
-            key={item.id}
-            onClick={this.handleItemSelection}
-            role="tab"
-            title={item.label}
-          >
-            <calcite-icon icon={isMatch ? ICONS.active : ICONS.inactive} scale="l" />
-          </button>
-        );
-      })}
-    </div>
-  );
+  renderPaginationItems = (): VNode => {
+    const { selectedIndex, maxItems, items, label, handleItemSelection } = this;
+    return (
+      <div aria-label={label} class={CSS.paginationItems} role="tablist">
+        {items.map((item, index) => {
+          const length = items.length;
+          const match = index === selectedIndex;
+          const first = index === 0;
+          const last = index === length - 1;
+          const endRangeStart = length - maxItems - 1;
+          const inStartRange = selectedIndex < maxItems;
+          const inEndRange = selectedIndex > endRangeStart;
+          const rangeStart = Math.max(0, selectedIndex - Math.floor(maxItems / 2));
+          const rangeEnd = Math.min(length, rangeStart + maxItems);
+          const low = inStartRange ? 0 : inEndRange ? endRangeStart : rangeStart;
+          const high = inStartRange ? maxItems + 1 : rangeEnd;
+          const isEdge = !first && !last && !match && (index === low - 1 || index === high);
+          const visible = match || (index <= high && index >= low - 1);
+          const overflowActive = length - 1 <= maxItems;
+          const icon = match ? ICONS.active : ICONS.inactive;
+
+          return (
+            <button
+              aria-controls={!match ? item.id : undefined}
+              aria-selected={toAriaBoolean(match)}
+              class={{
+                [CSS.paginationItem]: true,
+                [CSS.paginationItemIndividual]: true,
+                [CSS.paginationItemSelected]: match,
+                [CSS.paginationItemRangeEdge]: length - 1 > maxItems && isEdge,
+                [CSS.paginationItemOutOfRange]: !(overflowActive || visible),
+                [CSS.paginationItemVisible]: overflowActive || visible,
+              }}
+              data-index={index}
```

Comment:
Is the `data-index` used?

Reply:
Used here to determine the requested "item" https://github.com/Esri/calcite-design-system/blob/d53f47c0c86b6721a1b74e3c8ec69f20962d3543/packages/calcite-components/src/components/carousel/carousel.tsx#L453 - LMK if this should use a different pattern.

Reply:
No, this sounds good. 👍